spylib/token.py

```diff
@@ -0,0 +1,428 @@
+                'client': self.client,
+                'oauth_url': self.oauth_url,
+                'api_url': self.api_url,
+                'api_version': self.api_version,
+                'rest_bucket_max': self.rest_bucket_max,
+                'rest_leak_rate': self.rest_leak_rate,
+                'graphql_bucket_max': self.graphql_bucket_max,
+                'graphql_leak_rate': self.graphql_leak_rate,
+            }
+        )
+
+
+class OfflineTokenABC(Token, ABC):
+    """
+    Offline tokens are used for long term access, and do not have a set expiry.
+    """
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(**kwargs)
+
+    async def reset_token(self, client_id: str, client_secret: str, code: str):
+
+        token: OfflineTokenResponse = OfflineTokenResponse(
+            **await self._get_token(client_id, client_secret, code)
+        )
+        self.access_token = token.access_token
+        self.scope = token.scope.split(',')
+
+
+@dataclass
+class OnlineTokenABC(Token):
+    """
+    Online tokens are used to implement applications that are authenticated with
+    a specific user's credentials. These extend on the original token, adding
+    in a user, its scope and an expiry time.
+    """
+
+    associated_user_scope: Optional[List[str]]
+    associated_user: Optional[AssociatedUser]
+    expires_at: Optional[datetime]
+
+    def __init__(
+        self,
+        associated_user_scope: Optional[List[str]] = [],
+        associated_user: Optional[AssociatedUser] = None,
+        expires_in: int = 0,
+        **kwargs,
+    ) -> None:
+        super().__init__(
+            **dict(
+                **kwargs,
+                **{
+                    'associated_user_scope': associated_user_scope,
+                    'associated_user': associated_user,
+                    'expires_at': datetime.now() + timedelta(days=0, seconds=expires_in),
+                },
+            )
+        )
+
+    async def reset_token(self, client_id: str, client_secret: str, code: str):
+
+        token: OnlineTokenResponse = OnlineTokenResponse(
+            **await self._get_token(client_id, client_secret, code)
+        )
+        self.access_token = token.access_token
+        self.scope = token.scope.split(',')
+        self.associated_user = token.associated_user
+        self.associated_user_scope = token.associated_user_scope.split(',')
+        self.expires_at = datetime.now() + timedelta(days=0, seconds=token.expires_in)
+
+    async def save_token(self):
+        """
+        This method handles saving the token. By default this does nothing,
+        therefore the developer should override this.
+        """
+        pass
+
+    @classmethod
+    def load_token(cls, store_name: str, associated_user: str):
+        """
+        This method handles loading the token. By default this does nothing,
+        therefore the developer should override this.
+        """
+        pass
```

Comment:
Wait I just realized that you shouldn't have `load_token` in `Token` because the method arguments depends if you are in `OfflineToken` or `OnlineToken`.

Also this has to be async
```suggestion
    @classmethod
    async def load_token(cls, store_name: str, associated_user: str):
        """
        This method handles loading the token. By default this does nothing,
        therefore the developer should override this.
        """
        pass
```